src/NuGet.Core/NuGet.Commands/RestoreCommand/RequestFactory/DependencyGraphSpecRequestProvider.cs

```diff
@@ -94,6 +96,7 @@ private IReadOnlyList<RestoreSummaryRequest> GetRequestsFromItems(RestoreArgs re
                 {
                     IReadOnlyList<PackageSpec> closure = dgFile.GetClosure(projectNameToRestore);
                     DependencyGraphSpec projectDependencyGraphSpec = dgFile.CreateFromClosure(projectNameToRestore, closure);
+                    projectDependencyGraphSpec.SetProjectNameToHashCode(dictionary);
```

Comment:
public API design feedback: is there a reason to pass the dictionary in a method, rather than the `projectDependencyGraphSpec` constructor, and therefore through the `CreateFromClosure` method?

Are there any restore codepaths where the `DependencyGraphSpec` will not, or should not, have one of these caches?

Reply:
This makes the caching obvious and scopes it down to restore only and avoids unexpected behaviors. 

I thought about adding it in CreateFromClosure and it'd avoid the public method, but it'll lead to potential fragility. 
I don't feel too strongly about that though, as fragility is the name of the game in this PR as we basically assume no one will ever change the PackageSpec within the core of restore.

Reply:
You know, I think I prefer the new public API-less approach, even if CreateFromClosure becomes a bit funky. 

I'll document it and only do when the dg spec is read only.